09-optimization.Rmd

```diff
@@ -0,0 +1,668 @@
+```{r, include = FALSE}
+ottrpal::set_knitr_image_path()
+```
+
+# Optimization
+
+In our last chapter, we will explore ways to optimize our completed workflow.
+
+Generally, in computing, we are working with the following finite resources:
+
+-   **Computing time**: a task can take miliseconds to days to complete. We find ways to reduce our computing time so that we have results sooner.
+
+-   **Central Processing Unit (CPU) usage**: a task requires at least one CPU to be the "brain" of the computer to run computation, move data around, and access the memory (see below). Tasks can make use of multiple CPUs in parallel to divide the task into smaller tasks.
+
+-   **Memory usage (RAM)**: a task needs a place to store data while computing on the job. Reducing the memory usage can allow other tasks to be efficient.
+
+-   **Disk usage**: A task usually requires some input data and have an output, and needs disk storage to store the information when the task isn't running.
+
+Changing usage of any of these resources often affect other resources. For instance, there's often a tradeoff between computing time and memory usage. In order to be efficient with these finite resources, we will introduce some optimization methods that are common in building an efficient workflow, and provide an example on a simple optimization method, "Embarrassingly Parallel Scatter-Gather".
+
+## Common optimizing/parallelizing methods
+
+### Memory optimization
+
+To optimizie the amount of memory used for a task, one can profile the task to understand how much memory is being used and where the bottlenecks are. Most programming languages have a large selection of memory profilers to analyze the performance of your code, and they can be starting points to consider how to use less memory. When a bottleneck is found, one might consider to use a more efficient algorithm to reduce memory usage. Or, perhaps one might use more optimal data structures, such as databases or sparse data structures, that better fit the data type of your problem.
```

Comment:
I'd change the "one might" to the 2nd person (like "you might" in the following paragraphs).